### DISCUSSION TITLE
BulkAction : Generate PDF

### DISCUSSION THREAD

**Discussion by invaders-xx** [filamentphp/filament#2110] [Help]
I have a ressource with table BulkAction 
``  Tables\Actions\BulkAction::make('generateBill')->label(__('Generate Bill'))->action(function (Collection $records, $livewire) {
                        $livewire->generateBill($records);
                    })`` 
and in the list records page I have the public function 
``public function generateBill(Collection $records)
    {
        $pdfContent = Pdf::loadView('pdf.bill', ['records' => $records])->output();
        return response()->streamDownload(
            fn() => print($pdfContent),
            "invoice.pdf"
        );
    }``

 but I can't get the pdf file. Did I miss something ? (I use Barryvdh\DomPDF to generate PDF)

**Comment by danharrin** [comment]
Return the Livewire response from the action (as mentioned on Discord).

**Reply by photojo** [reply]
@invaders-xx 
Can you post your solution here, please. How did you solve the problem. I'm a Laravel beginner. I don't know, what danherrin mean with "as mentioned on Discord".
Thank you very much!

**Reply by photojo** [reply]
**Filament Table**
`
->action(function ($record) {
                        return response()->streamDownload(function () use ($record) {
                            echo  MyService::generatePDF($record['id']);
                        }, 'File_' . $record['firstname']. '.pdf');
                    }),
`

**MyService**

public static function generatePDF(string $id)
    {
        $user = user::where('id', $id)
            ->firstOrFail();

        $mpdf = new \Mpdf\Mpdf([
            'margin_left' => 0,

        ]);

        $mpdf->SHYlang = 'de';  $mpdf->SHYleftmin = 2;

        $tplId = $mpdf->ImportPage(1);
        $mpdf->UseTemplate($tplId);

        $html = view('pdf.template', [
            'user' => $user,
        ]);

        $mpdf->WriteHTML($html);

        $mpdf->Output();

        return $mpdf->Output('', \Mpdf\Output\Destination::STRING_RETURN);
    }


Everything works fine!

But I don't what to do in 

protected function getTableBulkActions(): array
    {
        return [
            BulkAction::make('File')
                ->action(fn(Collection $records)  ???

**Reply by photojo** [reply]
Ok. So long. The following works without errors, but I just get one pdf.

` BulkAction::make('File')
                ->action(function(Collection $records)
                {
                    foreach($records as $record)
                    {
                        return response()->streamDownload(function () use ($record) { echo MyService::generatePDF($record['id']); }, 'File_' . $record['firstname']. '.pdf'); })
                    }
                }
`

What to do?

**Reply by invaders-xx** [reply]
@photojo, you can create a zip archive containing all the PDF. I did a trick for that 
[Add bulk action to table to export selected record as JSON file within a ZIP archive](https://filamentphp.com/tricks/add-bulk-action-to-table-to-export-selected-record-as-json-file-within-a-zip-archive)

It's JSON related but you can do the same with PDF

**Reply by photojo** [reply]
Thank you for your quick response. I'll try to make with a zip-file.
Is it not possible to get single files without zipping them?

**Reply by photojo** [reply]
Your example works fine! But how can I put the pdfs in the archives?

**Reply by invaders-xx** [reply]
@photojo you have to dig into ZipArchive library and find how ;-)

**Reply by photojo** [reply]
Ok, I break off at this point. It's too difficult for me. ;-(

**Reply by invaders-xx** [reply]
Just generate your PDF's file and add those files in the ZipArchive using [addFile](https://www.php.net/manual/en/ziparchive.addfile.php)

**Comment by photojo** [comment]
Like
`
function (Collection $records) {
                        $archive = new \ZipArchive;
                        $archive->open('file.zip', \ZipArchive::CREATE | \ZipArchive::OVERWRITE)

                        foreach ($records as $record) {


                            **$file =  echo MyService::generatePDF($record['id']); }, 'File_' . $record['firstname']. '.pdf');**


                            $archive->addFile($file);
                        }
                        $archive->close();
                        return response()->download('file.zip');
                    }
`
?

Sorry, but I'm in despair. As I described a single table action works
`->action(function ($record) { 
return response()->streamDownload(function () use ($record) { 
echo MyService::generatePDF($record['id']); }, 'File_' . $record['firstname']. '.pdf'); }),`

I can't get these files in the zip-archive?

**Comment by photojo** [comment]
So, obviously I'm not good enough to do so. I give it up.
